pkg/cloud/services/loadbalancer/loadbalancer.go

```diff
@@ -166,6 +89,142 @@ func (s *Service) ReconcileLoadBalancer(openStackCluster *infrav1.OpenStackClust
 	return nil
 }
 
+func (s *Service) getOrCreateLoadBalancer(openStackCluster *infrav1.OpenStackCluster, loadBalancerName, subnetID string) (*loadbalancers.LoadBalancer, error) {
+	lb, err := checkIfLbExists(s.loadbalancerClient, loadBalancerName)
+	if err != nil {
+		return nil, err
+	}
+
+	if lb != nil {
+		return lb, nil
+	}
+
+	s.logger.Info(fmt.Sprintf("Creating load balancer in subnet: %q", subnetID), "name", loadBalancerName)
+	lbCreateOpts := loadbalancers.CreateOpts{
+		Name:        loadBalancerName,
+		VipSubnetID: subnetID,
+	}
+
+	lb, err = loadbalancers.Create(s.loadbalancerClient, lbCreateOpts).Extract()
+	if err != nil {
+		record.Warnf(openStackCluster, "FailedCreateLoadBalancer", "Failed to create load balancer %s: %v", loadBalancerName, err)
+		return nil, err
+	}
+
+	if err := waitForLoadBalancerActive(s.logger, s.loadbalancerClient, lb.ID); err != nil {
+		record.Warnf(openStackCluster, "FailedCreateLoadBalancer", "Failed to create load balancer %s: %v", loadBalancerName, err)
+		return nil, err
+	}
+
+	record.Eventf(openStackCluster, "SuccessfulCreateLoadBalancer", "Created load balancer %s with id %s", loadBalancerName, lb.ID)
+	return lb, nil
+}
+
+func (s *Service) getOrCreateListener(openStackCluster *infrav1.OpenStackCluster, listenerName, lbID string, port int) (*listeners.Listener, error) {
+	listener, err := checkIfListenerExists(s.loadbalancerClient, listenerName)
+	if err != nil {
+		return nil, err
+	}
+
+	if listener != nil {
+		return listener, nil
+	}
+
+	s.logger.Info("Creating load balancer listener", "name", listenerName, "lb-id", lbID)
+
+	listenerCreateOpts := listeners.CreateOpts{
+		Name:           listenerName,
+		Protocol:       "TCP",
+		ProtocolPort:   port,
+		LoadbalancerID: lbID,
+	}
+	listener, err = listeners.Create(s.loadbalancerClient, listenerCreateOpts).Extract()
+	if err != nil {
+		record.Warnf(openStackCluster, "FailedCreateListener", "Failed to create listener %s: %v", listenerName, err)
+		return nil, err
+	}
+
+	if err := waitForLoadBalancerActive(s.logger, s.loadbalancerClient, lbID); err != nil {
+		record.Warnf(openStackCluster, "FailedCreateListener", "Failed to create listener %s: %v", listenerName, err)
+		return nil, err
+	}
+
+	if err := waitForListener(s.logger, s.loadbalancerClient, listener.ID, "ACTIVE"); err != nil {
+		record.Warnf(openStackCluster, "FailedCreateListener", "Failed to create listener %s: %v", listenerName, err)
```

Comment:
should we also add the id here?

Reply:
I would prefer also adding something about that the wait for listener active failed